src/category_theory/cofiltered_system.lean

```diff
@@ -215,10 +294,68 @@ let ⟨i, f, h⟩ := F.is_mittag_leffler_iff_eventual_range.1 h j in
 by { rw [to_eventual_ranges_obj, h], apply_instance }
 
 /-- If `F` has all arrows surjective, then it "factors through a poset". -/
-lemma thin_diagram_of_surjective (Fsur : ∀ (i j : J) (f : i ⟶ j), (F.map f).surjective)
-  (i j) (f g : i ⟶ j) : F.map f = F.map g :=
+lemma thin_diagram_of_surjective (Fsur : ∀ ⦃i j : J⦄ (f : i ⟶ j), (F.map f).surjective)
+  {i j} (f g : i ⟶ j) : F.map f = F.map g :=
 let ⟨k, φ, hφ⟩ := cone_maps f g in
-(Fsur k i φ).injective_comp_right $ by simp_rw [← types_comp, ← F.map_comp, hφ]
+(Fsur φ).injective_comp_right $ by simp_rw [← types_comp, ← F.map_comp, hφ]
+
+lemma to_preimages_nonempty_of_surjective [hFn : ∀ (j : J), nonempty (F.obj j)]
+  (Fsur : ∀ ⦃i j : J⦄ (f : i ⟶ j), (F.map f).surjective)
+  (hs : s.nonempty) (j) : nonempty ((F.to_preimages s).obj j) :=
+begin
+  simp only [to_preimages_obj, nonempty_coe_sort, nonempty_Inter, mem_preimage],
+  obtain (h|⟨⟨ji⟩⟩) := is_empty_or_nonempty (j ⟶ i),
+  { exact ⟨(hFn j).some, λ ji, h.elim ji⟩, },
+  { obtain ⟨y, ys⟩ := hs,
+    obtain ⟨x, rfl⟩ := Fsur ji y,
+    exact ⟨x, λ ji', (F.thin_diagram_of_surjective Fsur ji' ji).symm ▸ ys⟩, },
+end
+
+lemma eval_section_injective_of_eventually_injective
+  {j} (Finj : ∀ i (f : i ⟶ j), (F.map f).injective) (i) (f : i ⟶ j) :
+  (λ s : F.sections, s.val j).injective :=
+begin
+  refine λ s₀ s₁ h, subtype.ext $ funext $ λ k, _,
+  obtain ⟨m, mi, mk, _⟩ := cone_objs i k,
+  dsimp at h,
+  rw [←s₀.prop (mi ≫ f), ←s₁.prop (mi ≫ f)] at h,
+  rw [←s₀.prop mk, ←s₁.prop mk],
+  refine congr_arg _ (Finj m (mi ≫ f) h),
+end
+
+section finite_cofiltered_system
+
+variables [∀ (j : J), nonempty (F.obj j)] [∀ (j : J), finite (F.obj j)]
+  (Fsur : ∀ ⦃i j : J⦄ (f :i ⟶ j), (F.map f).surjective)
+
+include Fsur
+lemma eval_section_surjective_of_surjective (i : J) :
+  (λ s : F.sections, s.val i).surjective := λ x,
+begin
+  let s : set (F.obj i) := {x},
+  haveI := F.to_preimages_nonempty_of_surjective s Fsur (singleton_nonempty x),
+  obtain ⟨sec, h⟩ := nonempty_sections_of_finite_cofiltered_system (F.to_preimages s),
+  refine ⟨⟨λ j, (sec j).val, λ j k jk, by simpa [subtype.ext_iff] using h jk⟩, _⟩,
+  { have := (sec i).prop,
+    simp only [mem_Inter, mem_preimage, mem_singleton_iff] at this,
+    replace this := this (𝟙 i), rwa [map_id_apply] at this, },
+end
+
+lemma eventually_injective [nonempty J] [finite F.sections] :
+  ∃ j, ∀ i (f : i ⟶ j), (F.map f).injective :=
+begin
+  haveI : ∀ j, fintype (F.obj j) := λ j, fintype.of_finite (F.obj j),
+  haveI : fintype F.sections := fintype.of_finite F.sections,
+  have card_le : ∀ j, fintype.card (F.obj j) ≤ fintype.card F.sections :=
+    λ j, fintype.card_le_of_surjective _ (F.eval_section_surjective_of_surjective Fsur j),
+  let fn := λ j, fintype.card F.sections - fintype.card (F.obj j),
```

Comment:
Using `nat.card` instead would allow us to remove the `haveI` and make the proof even shorter, if we're okay with adding an import.

Reply:
I don't mind either: the proof looks short enough as-is, but I'm curious how you'd do it with `nat.card` given that I don't see any equivalent to `fintype.card_le_of_surjective` and `fintype.bijective_iff_surjective_and_card`.

Reply:
I think it suffices to use
[finite.card_le_of_surjective'](https://leanprover-community.github.io/mathlib_docs/data/finite/card.html#finite.card_le_of_surjective')
[function.surjective.injective_of_fintype](https://leanprover-community.github.io/mathlib_docs/data/fintype/card.html#function.surjective.injective_of_fintype)
[finite.card_eq](https://leanprover-community.github.io/mathlib_docs/data/finite/card.html#finite.card_eq)

Reply:
Probably no point to import cardinals though. So

maintainer merge